pkg/lib/indexer/indexer.go

```diff
@@ -283,3 +292,182 @@ func write(dockerfileText, outDockerfile string, logger *logrus.Entry) error {
 
 	return nil
 }
+
+// ExportFromIndexRequest defines the parameters to send to the ExportFromIndex API
+type ExportFromIndexRequest struct {
+	Index         string
+	Operator      string
+	DownloadPath  string
+	ContainerTool string
+}
+
+// ExportFromIndex is an aggregate API used to specify operators from
+// an index image
+func (i ImageIndexer) ExportFromIndex(request ExportFromIndexRequest) error {
+	databaseFile := defaultDatabaseFile
+
+	// set a temp directory
+	workingDir, err := ioutil.TempDir("./", "index_tmp")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(workingDir)
+
+	// Pull the fromIndex
+	if request.Index != "" {
+		i.Logger.Infof("Pulling previous image %s to get metadata", request.Index)
+
+		// Get the old index image's dbLocationLabel to find this path
+		labels, err := i.LabelReader.GetLabelsFromImage(request.Index)
+		if err != nil {
+			return err
+		}
+		if dbLocation, ok := labels[containertools.DbLocationLabel]; ok {
+			i.Logger.Infof("Previous db location %s", dbLocation)
+
+			// extract the database to the file
+			err = i.ImageReader.GetImageData(request.Index, workingDir)
+			if err != nil {
+				return err
+			}
+
+			databaseFile = path.Join(workingDir, dbLocation)
+		}
+	} else {
+		databaseFile = path.Join(workingDir, databaseFile)
+	}
+
+	bundles, err := getBundlesToExport(databaseFile, request.Operator)
+	if err != nil {
+		return err
+	}
+	i.Logger.Infof("Preparing to pull bundles %+q", bundles)
+
+	// Creating downloadPath dir
+	if _, err := os.Stat(request.DownloadPath); err != nil {
+		if os.IsNotExist(err) {
+			err = os.Mkdir(request.DownloadPath, 0777)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	var errs []error
+
+	db, err := sql.Open("sqlite3", databaseFile)
+	if err != nil {
+		return err
+	}
+
+	dbQuerier := sqlite.NewSQLLiteQuerierFromDb(db)
+	if err != nil {
+		return err
+	}
+
+	for _, bundleImage := range bundles {
+		// try to name the folder
+		folderName, err := dbQuerier.GetBundleVersion(context.TODO(), bundleImage)
+		if err != nil {
+			return err
+		}
+		if folderName == "" {
+			// operator-registry does not care about the folder name
+			folderName = bundleImage
+		}
+		exporter := bundle.NewSQLExporterForBundle(bundleImage, filepath.Join(request.DownloadPath, folderName), request.ContainerTool)
+		if err := exporter.Export(); err != nil {
+			err = fmt.Errorf("error exporting bundle from image: %s", err)
+			errs = append(errs, err)
+		}
+	}
+	err = db.Close()
+	if err != nil {
+		errs = append(errs, err)
+		return utilerrors.NewAggregate(errs)
+	}
+
+	err = generatePackageYaml(databaseFile, request.Operator, request.DownloadPath)
+	if err != nil {
+		errs = append(errs, err)
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+func getBundlesToExport(databaseFile, packageName string) ([]string, error) {
```

Comment:
is there a reason that these helpers are opening/closing the file? can we pass in the db reference or the querier instead of the filename?

Reply:
That's what I did at first but changed it to make testing easier

Reply:
Can you explain how this simplifies testing?